Bluetooth: Controller: Add support to transmit Constant Tone Extension in periodic advertising

ppryga-nordic opened the pull request:
The main goal of the PR is to add support to transmit CTE in periodic advertising chains.
To achieve that following code functionalities were added or removed:
- enabled connectionless CTE feature,
- enabled reporting about received CTE,
- added `extra_data` storage for CTE additional radio configuration (more on that in commit message: https://github.com/zephyrproject-rtos/zephyr/commit/d9d6844e16c63e1042e2520564288b847c6021eb)
- extended nRF radio API to add possibility to configure Direction Finding Extension 
- added support to handle CTE transmission in LLL
- re-worked handling of extended advertising PDU for periodic advertising ULL
- added support to enable transmission of CTE in ULL.

There are missing unit tests for the functionality.
It will be provided, hopefully next week.

ppryga-nordic committed:
Bluetooth: controller: ll_feat: Add CTE connectionless TX feat.

Add Directin Finding related features:
- transmission of connectionless CTE,
- antenna switching during CTE transmission (AoD),
to list of features supported by controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: hci: Enable reporting of CTE reception

Enable reporting CTE in periodic advertising reports send to host.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: hci: Fix wrong values of cte type constants

Wrong values were used for CTE type constants. They were
update to comply with BT 5.1 spec.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Add extra data storage for ext. adv. configuration

The storage for extra data is required for implementation of
CTE transmission with periodic advertising.

Data required to transmit CTE correctly are compound of two parts:
- PDU field CTEInfo
- radio configuration to transmit actual constant tone at the end
  of PDU.

Extra data is a storage required for radio configuration data.
Nevertheless it must be in compliance with content of CTEInfo field.
Because of that extra data is stored as part of lll_adv_pdu and
is double buffered like PDU memory.

Bluetooth 5.1 spec. allows to enable or disable CTE TX
and change CTE TX parameters when periodic advertising is
enabled. Besides that CTE TX settings may be set before periodic
advertising parameters are set. In such situation ll_adv_sync_set
may be not yet created.
To overcome these constraints ULL should store CTE TX
settings and forward them to LLL only when CTE TX is enabled.
Because of above reasons ULL stores CTE TX settings in ll_adv_set.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: radio: Add DF configuration to enable CTE TX

Add set of functions that will make possible to configure
radio Direction Finding Extension to transmit CTE for periodic
advertising.

Some of the new Radio API functions are provided as separate
functions changing the same Radio peripheral registers, e.g.
radio_df_mode_set_aoa, radio_df_mode_set_aod. This is done on
purpose and is related with lack of DFE in nrf52_bsim.
To avoid use of conditionally compiled constants to represent
e.g. CTE mode; separate functions were introduced.
Thanks to that DF unit tests are able to compile successfully
without changes in nrf52_bsim platform. Also if DFE is added
to nrf52_bsim there is no need to change the code until it is
desired.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: enable TX of CTE with per. adv. PDU

Enable transmission of CTE with periodic advertising PDU.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: Add CTE to per. adv. PDU

Add CTE to extended header of periodic advertising PDU.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: ULL: enable TX of CTE with per. adv. PDU

Enable transmission of CTE with periodic advertising PDU.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Changes provided:
- rebase to fix merge conflict
- changed implementation of DF Radio peripheral API. This is required to make possible to build DF code for unit tests and nrf52_bsim.

ppryga commented:
Fixed error in UT build.

ppryga-nordic committed:
tests: Bluetooth: df: Move common code to separate source file

Move common code that may be used by different test units
from main.c to common.c. Main.c should be clean and responsible
for unit tests executioon only. Common functionalities should be
stored in separate file and provided to test source files by
appropriate header file.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: add UT for enable CTE for periodic advertising

Adds unit tests for verification of enable/disable CTE for periodic
advertiging in controller.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Changes added:
- implemented UT for CTE enable/disable in periodic advertising
- fixed issues in builds with disabled DF (all UT executed by CI are passing now)
- fixed extra_data memory leak when allocate new PDU memory without extra_data requested.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -56,170 +59,234 @@ static void ticker_op_cb(uint32_t status, void *param);
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+static uint8_t adv_sync_hdr_set_clear(struct lll_adv_sync *lll_sync,
```
Move this static function after the definition of `sync_remove` in this file.

ppryga replied:
Code moves according to suggestion

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -206,11 +252,48 @@ int lll_adv_data_release(struct lll_adv_pdu *pdu)
 	return 0;
 }
 
+static struct pdu_adv *adv_pdu_allocate(struct lll_adv_pdu *pdu, uint8_t last)
```
Move all static functions in the logical order of execution after all the non-static functions in this file.

ppryga replied:
Code moved according to suggestion.

What do you thing about extract of functions responsible for PDU memory handling into separate file e.g. lll_adv_pdu.c?
The code servers now legacy, extended and periodic advertising. It may improve readability of the lll_adv.c source code.

cvinayak replied:
No objections to  have pdu operations in separate file, but do it in a separate PR? (just being cautious about merge conflicts to other team members working on same file)

ppryga replied:
Yes, separate PR is better option. Lets postpone it until 2.5 is released, then we can schedule it when all team members are ok to make this change.

cvinayak reviewed (changes requested):
Reviewed for controller implementation design and style, only minor comments to move function inside the files.
Did not test on target or review the tests, quick-ack on these.

ppryga commented:
Implemented changes requested by @cvinayak

ppryga commented:
@kruithofa @thoh-ot @Emil-Gydesen-Bose could you take a look on the PR?

ppryga-nordic committed:
Bluetooth: controller: Add release of periodic advertising sync set

Add release of memory acquired for periodic advetising sync. set
when:
- advertising set is removed
- advertising parameters are changed and past PDU type was
  extended advertising PDU

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Add release of memory acquired for periodic advertising sync. set
    when:
    - advertising set is removed
    - advertising parameters are changed and past PDU type was
      extended advertising PDU

ppryga commented:
Updated commit message.

Thalley commented on tests/bluetooth/df/src/common.c:
```diff
@@ -53,3 +61,62 @@ void common_delete_adv_set(void)
+	struct net_buf *buf;
+	int err;
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_CL_CTE_TX_PARAMS,
+				sizeof(*cp) + ARRAY_SIZE(ant_ids));
+	zassert_not_null(buf, "Failed to create HCI cmd object");
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->handle = g_adv->handle;
+	cp->cte_len = g_cte_len;
+	cp->cte_type = BT_HCI_LE_AOD_CTE_2US;
+	cp->cte_count = 5;
+
+	dest_ant_ids = net_buf_add(buf, ARRAY_SIZE(ant_ids));
+	memcpy(dest_ant_ids, ant_ids, ARRAY_SIZE(ant_ids));
+
+	cp->switch_pattern_len = ARRAY_SIZE(ant_ids);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_CL_CTE_TX_PARAMS,
+				   buf, NULL);
```
Shouldn't something like this be implemented in hci_core.c and called from here instead?

ppryga replied:
There is a function that does the same thing `hci_df_set_conn_cte_tx_param` but it does additional parameters validation that is not necessary here and the function is static (implemented in `direciton.c`). 
There is a public API function `bt_df_set_adv_cte_tx_param` (implemented in the same file) but this one uses atomics etc. to check if valid sequence of function calls is kept. 

If I go with the `bt_df_set_adv_cte_tx_param` then the tests will start to look like sample application not unit test of controller functionality.

Thalley reviewed (commented):
I'm not too familiar with CTE as a whole, so I kept my review to the code itself, and should be considered a superficial review, but had a few comments

ppryga commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -171,6 +171,15 @@ config BT_CTLR_ADV_DATA_BUF_MAX
 	  Maximum number of buffered Advertising Data payload across enabled
 	  advertising sets.
 
+config BT_CTLR_ADV_EXT_PDU_EXTRA_DATA
+	bool "Storage for extra data added to extended advertising PDU structure"
+	depends on BT_CTLR_ADV_EXT
+	help
+	  Enables possiblity to add extra data to extended advertising PDU
+	  structure that are used for preparation of advertising PDU. This gives
+	  possiblity to send additional information from ULL to LLL when it
+	  is already advertising.
```
I've named the config as well as advertising PDU data structure member `extra_data` to make it general. At the moment there is only one use-case for it, to provide additional configuration for Radio that must be in sync with content of advertising PDU (more precisely CTEInfo field of ext. adv. header).  I wanted to be more generic, because it is just a memory transport information between ULL and LLL. It is no a part of actual PDU and may be used to provide any other data required by LLL. In future it may be used for any other purpose.

Name of a config is tightly related with a data structure where it is added. 

It is a good idea to make the config hidden, because it should not be altered by end user. It serves other functionalities.

Referring to description, have you got any suggestions how to make it easier to understand?

Thalley replied:
If we make it hidden, I don't think we need to improve the text, as user's won't need to worry about it, and we can leave it as is. 

Regarding appending "_MEMORY" I don't think that really helps much either :)

ppryga replied:
I've made the config to be hidden. Also I've decided to change the name with _MEMORY. If anyone has a better proposal for the name it is welcome. Also I've reworded description, hope it is better to understand the purpose of the config and memory related with it.

I decided to do changes. besides hide the option because it is still a documentation for other developers.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -235,58 +309,218 @@ struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx)
+		void *p;
+
+		if (!MFIFO_ENQUEUE_IDX_GET(pdu_free, &free_idx)) {
+			LL_ASSERT(false);
+
+			return NULL;
```
Well, this is a question to @cvinayak. I didn't add this function. Please take a look on master: https://github.com/zephyrproject-rtos/zephyr/blob/4d4a636555cdec8785652e016d974667be7c201b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L272

cvinayak replied:
Its in the upstream code. I agree with @Thalley. @ppryga if its not too much, could you remove this `LL_ASSERT` and move to callers of `lll_adv_pdu_latest_get`

ppryga replied:
@cvinayak sure I'll remove it

ppryga replied:
Guys, I suggest to go with remove of return NULL. 
Please take a look e.g. here: https://github.com/zephyrproject-rtos/zephyr/blob/4d4a636555cdec8785652e016d974667be7c201b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L910-L912

chan_prepare function hasn't got any way to handle it. If I add LL_ASSERT(!pdu); here then a code size increases. It will increase in every place I'll insert LL_ASSER. (I hope lto can do something about it). Besides that there is maintenance cost.

If I leave LL_ASSERT in `lll_adv_pdu_latest_get` then code size will not rise and there will be less maintenance.

ppryga replied:
I've removed `return NULL` following reasoning from my last comment.

Thalley replied:
Correct me if I'm wrong, but I don't think we should be concerned about code sizes from using LL_ASSERT: My understanding is that these would be compiled out for production builds.

ppryga replied:
Yes, that is true. My bad, sorry. But... in such case, does it make sense to use LL_ASSERT in caller of the function? 
At the end of the day it will be removed. 

In my humble opinion the issue is still in the `lll_adv_pdu_latest_get` function (and other like this). 
If something very wrong had happen, then there wouldn't be a space to release past PDU (it would mean that we had a dangling pointer or more PDU than `pdu_free`  is able to store).
In such situation I would put here unconditional exception... 
If we leave it and return NULL then caller have to throw exception because in most cases there is no fall-back path.

ppryga replied:
There are other functions that have similar issue e.g. `lll_adv_pdu_alloc`: https://github.com/zephyrproject-rtos/zephyr/blob/3cf64b17c35ea12c6d3fa49086d09374acc9149b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c#L261-L269

Here we also have LL_ASSERT(p), but if `p` is `NULL` (there is no more free memory buffers for PDU) we are in a forest and cannot recover from that. 

What more it is not verified if the function returned NULL due to believe that it may not happen e.g. https://github.com/zephyrproject-rtos/zephyr/blob/3cf64b17c35ea12c6d3fa49086d09374acc9149b/subsys/bluetooth/controller/ll_sw/ull_adv.c#L1503-L1514

@Thalley you have opened pandora box ;-) 

@cvinayak What do you suggest to do with it?

Thalley replied:
One reason to have the LL_ASSERT in the caller, rather than the callee, is that for debugging it makes it easier to figure out the context of the issue. 

It's very valid to find examples of where it's done like you added: Whatever the resolution is here, then perhaps that should be applied to other places afterwards.

ppryga replied:
I've added changes as requested by @Thalley and @cvinayak. All changes are done as separate commit in PR.

I didn't add verification of return value for `lll_adv_pdu_alloc`. That would introduce more changes in callers and callers of callers... dure to use of `lll_adv_pdu_alloc` in functions that return `void`.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -111,70 +212,59 @@ uint8_t ll_adv_sync_param_set(uint8_t handle, uint16_t interval, uint16_t flags)
+	/* TODO: handle other op values */
+	if (op != BT_HCI_LE_EXT_ADV_OP_COMPLETE_DATA &&
+	    op != BT_HCI_LE_EXT_ADV_OP_UNCHANGED_DATA) {
+		/* FIXME: error code */
```
I believe so, this line was also not altered by me in this commit. Git marks it as added because it was moved.

cvinayak replied:
still valid, fragments are not supported yet.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -508,6 +517,25 @@ void ull_adv_sync_offset_get(struct ll_adv_set *adv)
 }
 
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ADV_CTE_TX)
+void ull_adv_sync_update(struct ll_adv_sync_set *sync, bool evt_extend,
+			 uint32_t update_us)
```
`evt_extend` informs if duration of an event should be extended of shortened. The other option is to use `int32_t` and if `update_us < 0` use it to shorten event duration.

Thalley replied:
Ah, this is my bad. I missed the `!` on the second line. 

I'm not too familiar with the ticker function, but I see that it uses both a plus and a negative argument (can both be set? Doesn't seem be any check in the implementation of the ticker). Perhaps this function should have both as well (i.e. a plus and a minus argument) to better match the ticker_update function (as this is basically just a wrapper around that). 

If both cannot (or should never) be set at the same time (i.e. a plus and a minus), then I like the idea of the signed parameter instead, giving that it can still handle the needed value range.

ppryga replied:
I'll go with two `uint32_t` arguments `slot_plus_us` and `slot_minus_us` because:
- this is more optimal way, less instructions are generated due to lack of conditional statements (I should have gone for that in first place),
- I want to avoid types conversion because it will introduce unnecessary instructions

ppryga replied:
Changed according the above description

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -586,6 +614,236 @@ static inline uint8_t sync_remove(struct ll_adv_sync_set *sync,
+ *
+ * @return Zero in case of success, other value in case of failure.
+ */
+static uint8_t adv_sync_hdr_set_clear(struct lll_adv_sync *lll_sync,
+				      struct pdu_adv *ter_pdu_prev,
+				      struct pdu_adv *ter_pdu,
+				      uint16_t hdr_add_fields,
+				      uint16_t hdr_rem_fields,
+				      void *value)
```
I did it this way because I wanted to avoid commit altering changes in previous commit.
May I consider the suggestion about separate commit as a hint for future?

Thalley replied:
Sure - It'll be a pain to separate them now :D

ppryga commented on subsys/bluetooth/controller/include/ll_feat.h:
```diff
@@ -111,6 +111,20 @@
 
+#if IS_ENABLED(CONFIG_BT_CTLR_DF) && IS_ENABLED(CONFIG_BT_CTLR_DF_ADV_CTE_TX)
+	#define LL_FEAT_BIT_CONNECTIONLESS_CTE_TX \
+		BIT64(BT_LE_FEAT_BIT_CONNECTIONLESS_CTE_TX)
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -47,6 +49,50 @@ uint8_t lll_df_ant_num_get(void)
 
+/* @brief Function initializes transmission of Constant Tone Extension.
+ *
+ * @param[in] type      Type of CTE: AoA, AoD 1us, AoD 2us
```
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -586,6 +614,236 @@ static inline uint8_t sync_remove(struct ll_adv_sync_set *sync,
+
+	/* No ADI in AUX_SYNC_IND*/
+
+	/* TODO: CTEInfo */
```
Removed TODO

ppryga-nordic committed:
Bluetooth: controller: Remove assert in functions that get latest adv. PDU

Remove LL_ASSERT from functions that return latest advertising PDU.
The LL_ASSERT was raised in situation that there is no memory to
store unused PDUs memory in a pdu_free queue or extra_data_free queue
Those functions return NULL in such sitation. The returned value is
verified by callers by LL_ASSERT. That gives better context if lack
of memory issue issue occurs.

Besides that there was removed a LL_ASSERT from lll_adv_pdu_and_extra_-
data_alloc. The reasons is the same as above, to give better context
when the lack of memory issue occurs.
This function is used in ULL context (ll_adv_sync_ad_data_set).
If it returns NULL the caller will return BT_HCI_ERR_MEM_CAPACITY_-
EXCEEDED to Host.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Implemented changes requested by @Thalley .
Fixed problem in tests, related with change _cpu_dsb() to _cpu_dmb().

Thalley reviewed (approved):
LGTM. Probably want to get an OK from @cvinayak regarding the LL_ASSERT.

ppryga commented:
Fixed an issue related with switch pattern setting for guard period (period before start of CTE (reference period). 
The guard period is not mentioned in BT 5.1 specification so it is not part of antenna switch pattern list provided from Host to controller. To avoid issues in compliance reference antenna is set as guard antenna also. Due to that number of supported switch patterns for Nordic devices is max 39 (former value was 40).

cvinayak reviewed (approved):
> 
> 
> LGTM. Probably want to get an OK from @cvinayak regarding the LL_ASSERT.

LL_ASSERT changes LGTM.
